engine/runtime/src/main/java/org/enso/interpreter/runtime/data/ArrayProxy.java

```diff
@@ -0,0 +1,74 @@
+package org.enso.interpreter.runtime.data;
+
+import com.oracle.truffle.api.CompilerDirectives;
+import com.oracle.truffle.api.interop.ArityException;
+import com.oracle.truffle.api.interop.InteropLibrary;
+import com.oracle.truffle.api.interop.InvalidArrayIndexException;
+import com.oracle.truffle.api.interop.TruffleObject;
+import com.oracle.truffle.api.interop.UnsupportedMessageException;
+import com.oracle.truffle.api.interop.UnsupportedTypeException;
+import com.oracle.truffle.api.library.CachedLibrary;
+import com.oracle.truffle.api.library.ExportLibrary;
+import com.oracle.truffle.api.library.ExportMessage;
+import org.enso.interpreter.dsl.Builtin;
+import org.enso.interpreter.runtime.Context;
+import org.enso.interpreter.runtime.error.PanicException;
+import org.enso.interpreter.runtime.library.dispatch.TypesLibrary;
+
+/**
+ * A wrapper that allows to turn an Enso callback providing elements into a polyglot Array.
+ *
+ * <p>This allows creation of arrays (and with them, vectors) using non-standard storage - for
+ * example exposing rows of a Table without copying any data.
+ */
+@ExportLibrary(InteropLibrary.class)
+@ExportLibrary(TypesLibrary.class)
+@Builtin(pkg = "immutable", stdlibName = "Standard.Base.Data.Array_Proxy.Array_Proxy")
+public final class ArrayProxy implements TruffleObject {
+  private final long length;
+  private final Object at;
+
+  @Builtin.Method(description = "Creates an array backed by a proxy object.")
+  public ArrayProxy(long length, Object at) {
+    if (CompilerDirectives.inInterpreter()) {
+      InteropLibrary interop = InteropLibrary.getUncached();
+      if (!interop.isExecutable(at)) {
+        throw new PanicException(
+            Context.get(interop).getBuiltins().error().makeTypeError("Function", at, "at"),
+            interop);
+      }
+    }
+
+    this.length = length;
+    this.at = at;
+  }
+
+  @ExportMessage
+  public boolean hasArrayElements() {
+    return true;
+  }
+
+  @ExportMessage
+  public long getArraySize() {
+    return length;
+  }
+
+  @ExportMessage
+  boolean isArrayElementReadable(long index) {
+    return index < length && index >= 0;
+  }
+
+  @ExportMessage
+  public Object readArrayElement(long index, @CachedLibrary(limit = "3") InteropLibrary interop)
+      throws UnsupportedMessageException, InvalidArrayIndexException {
+    if (index >= length || index < 0) {
```

Comment:
I don't think that's correct or even necessary? If it is outside of the range but you tried to access it you should get UnsupportedMessageException?  Right @JaroslavTulach ?

Reply:
It is necessary to get the semantics that one would normally expect.

If I remove it, the following test:
```
        Test.specify "should correctly delegate to the callback" <|
            arr = Array_Proxy.new 3 (ix -> ix + 10)
            arr.length . should_equal 3
            arr.at 0 . should_equal 10
            arr.at 1 . should_equal 11
            arr.at 2 . should_equal 12
            arr.at 3 . should_fail_with Index_Out_Of_Bounds_Error_Data
```
will fail like so:
```
Array_Proxy:  [168ms]
    - [FAILED] should correctly delegate to the callback [103ms]
        Reason: Expected an error Constructor<Index_Out_Of_Bounds_Error_Data> but no error occurred, instead got: 13 (at /home/radeusgd/NBO/enso/test/Tests/src/Data/Array_Proxy_Spec.enso:20:13-70).
```